server/src/main/java/org/elasticsearch/snapshots/SnapshotsService.java

```diff
@@ -2464,203 +2465,255 @@ public boolean assertAllListenersResolved() {
      * Package private to allow for tests.
      */
     static final ClusterStateTaskExecutor<ShardSnapshotUpdate> SHARD_STATE_EXECUTOR = (currentState, tasks) -> {
-        int changedCount = 0;
-        int startedCount = 0;
+        final SnapshotShardsUpdateContext update = new SnapshotShardsUpdateContext(currentState, tasks);
+        for (SnapshotsInProgress.Entry entry : currentState.custom(SnapshotsInProgress.TYPE, SnapshotsInProgress.EMPTY).entries()) {
+            update.applyToEntry(entry);
+        }
+        return ClusterStateTaskExecutor.ClusterTasksResult.<ShardSnapshotUpdate>builder().successes(tasks).build(update.updatedState());
+    };
+
+    private static boolean isQueued(@Nullable ShardSnapshotStatus status) {
+        return status != null && status.state() == ShardState.QUEUED;
+    }
+
+    /**
+     * State machine for updating existing {@link SnapshotsInProgress.Entry} by applying a given list of {@link ShardSnapshotUpdate} to
+     * them.
+     */
+    private static final class SnapshotShardsUpdateContext {
+
+        // number of updated shard snapshot states as a result of applying tasks to the snapshot entries seen so far
+        private int changedCount = 0;
+
+        // number of started tasks as a result of applying tasks to the snapshot entries seen so far
+        private int startedCount = 0;
+
+        // current cluster state
+        private final ClusterState currentState;
+
+        // snapshot entries computed by applying tasks to existing snapshot entries
         final List<SnapshotsInProgress.Entry> entries = new ArrayList<>();
-        final String localNodeId = currentState.nodes().getLocalNodeId();
-        // Tasks to check for updates for running snapshots.
-        final List<ShardSnapshotUpdate> unconsumedTasks = new ArrayList<>(tasks);
-        // Tasks that were used to complete an existing in-progress shard snapshot
+
+        // tasks outstanding to be applied to existing snapshot entries
+        final List<ShardSnapshotUpdate> unconsumedTasks;
```

Comment:
Could we make the naming more consistent, we're calling these things both "tasks" and "updates", I think it'd be better to stick closer to the class naming. Also `executedTasks` and really everywhere else where there's just one object of a particular class in scope.

Reply:
Tried to adjust the wording to `update` instead of `task` in all spots that I could find :)